First working version of the alternative JetMatching for Pythia8.3 and MG5aMCNLO

Cvico opened the pull request:
#### PR description:
This PR includes a plugin that allows to run an alternative Pythia8 FxFx jet matching prescription. The plugin has been implemented using the `PluginFactory` and the code has been directly taken from the new `JetMaching.h` file that can be found [here](http://amcatnlo.web.cern.ch/amcatnlo/FxFx_merging.htm). As stated in the link, this new jet matching algorithm only works in Pythia8.3 and beyond.

New versions of Madgraph5_aMC@NLO (v3.3.0 and beyond) introduce a new clustering procedure that distinguishes between jets that have been generated in a QCD or EWK vertex. This new jet matching prescription is required so the proper treatment of  jets is also accounted for in the parton shower (PS) and matrix element matching step.

An example of its usage can be found in [this fragment example](https://github.com/Cvico/cmssw/blob/JetMatchingEWKFxFx/GeneratorInterface/Pythia8Interface/python/FxFxEWK.py). Basically, the user needs to apply two modifications to the common fragments used for FxFx merging:
* Turn off Pythia8 default matching so new matchings can be add applied UserHooks. This can be done by adding the following line into the process parameters of the fragment:
`'JetMatching:doVeto = off',  # Turn OFF Pythia JetMatching`
* Include the new matching as a UserHook via the UserCustmization method.
```
    UserCustomization = cms.VPSet(
           cms.PSet(pluginName = cms.string("JetMatchingEWKFxFx"))
           ),
```

#### PR validation:

- Tests using the [local config file](https://github.com/Cvico/cmssw/blob/JetMatchingEWKFxFx/GeneratorInterface/Pythia8Interface/python/FxFxEWK.py) show that the JetMatchingEWKFxFx plugin properly matches between Pythia8.3 and MG33X gridpacks.
    - A set of preliminary results for ttW production can be found [here](https://indico.cern.ch/event/1192555/contributions/5086026/attachments/2525008/4342613/ttWnewFxFxSample_CarlosVico_TMG10oct.pdf).
#### Backporting
If merged, we aim for backporting this PR to 12_4_X for Run3 MC production.

Cvico committed:
First working version of the alternative JetMathing for Pythia8.3 and MG5

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-39996/32930

- This PR adds an extra 28KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-checks**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-39996/32930/code-checks.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-39996/32930/code-checks.patch | patch -p1`
You can also run `scram build code-checks` to apply code checks directly

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-39996/32930/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-39996/32930/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

Cvico committed:
Applied code-checks and formats for CMSSW coding rules compliance

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-39996/32936

- This PR adds an extra 28KB to repository

cmsbuild commented:
A new Pull Request was created by  @Cvico (Carlos Vico Villalba) for master.

It involves the following packages:

- GeneratorInterface/Pythia8Interface (**generators**)


@SiewYan, @mkirsano, @Saptaparna, @cmsbuild, @alberto-sanchez, @menglu21, @GurpreetSinghChahal can you please review it and eventually sign? Thanks.
@alberto-sanchez, @mkirsano this is something you requested to watch as well.
@perrotta, @dpiparo, @rappoccio you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

SiewYan commented:
please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-f6f25e/28854/summary.html
**COMMIT**: 637b23628a79719bd6a44c07fa85c00d18aefdac
**CMSSW**: CMSSW_12_6_X_2022-11-07-2300/el8_amd64_gcc10
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/39996/28854/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 10 differences found in the comparisons
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3416402
 - DQMHistoTests: Total failures: 12
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3416368
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 206 log files, 48 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_6_X_2022-11-07-2300+f6f25e/53960/triggerResults): no differences found

Cvico commented:
Dear reviewers,

I was wondering if there's anything more to do from my size for the final inclusion of 
this pull request. I see that code-checks have been ok for quite some time, but I don't
know if there's anything more to do. 

Thank you for your help,
Carlos Vico

menglu21 commented:
please test

menglu21 commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @perrotta, @dpiparo, @rappoccio (and backports should be raised in the release meeting by the corresponding L2)

perrotta commented on GeneratorInterface/Pythia8Interface/plugins/JetMatchingEWKFxFx.cc:
```diff
@@ -0,0 +1,997 @@
+  eventProcess.init("(eventProcess)", particleDataPtr);
+  workEventJet.init("(workEventJet)", particleDataPtr);
+
+  // Print information
```
`cout` is not thread sale: please think about protecting it with the `MATCHINGDEBUG` directive, as you do in all other cases, or move to some MessageLogger report instead

perrotta commented:
@Cvico I miss where the `JetMatching:doVeto` flag isread and then used in the code: is it outside your `JetMatchingEWKFxFx` class?

perrotta commented:
> This PR includes a plugin that allows to run an alternative Pythia8 FxFx jet matching prescription. The plugin has been implemented using the `PluginFactory` and the code has been directly taken from the new `JetMaching.h` file that can be found [here](http://amcatnlo.web.cern.ch/amcatnlo/FxFx_merging.html). As stated in the link, this new jet matching algorithm only works in Pythia8.3 and beyond.

Please notice that the link that you provided does not work: please fix it

Cvico committed:
Put the debug printout within a MATCHINGDEBUG block

Cvico commented:
Hi @perrotta 

Thank you for your prompt review. For some reason I can't reply to all the comments separately, so I'll address them
in here.

[comment 1.](https://github.com/cms-sw/cmssw/pull/39996#discussion_r1046843443) I guess here you are refering to the Madgraph matching parameters debug text? That piece of code is inherited from the original JetMatching.h file from pythia. Sure I can put it in a matchingdebug block, I don't think that's going to be a problem. I've updated the code according to this suggestion.

[comment 2.](https://github.com/cms-sw/cmssw/pull/39996#issuecomment-1348003053) So as you can see from my config file, I set the JetMatching::doVeto option to OFF in the fragment ([this line](https://github.com/Cvico/cmssw/blob/JetMatchingEWKFxFx/GeneratorInterface/Pythia8Interface/python/FxFxEWK.py#L33)). That tells pythia not to run its own JetMatching algorithm, but rather to expect an external matching algorithm to be called. So that option only affects the JetMatching from within pythia, and not this new matching that I implemented (that's why is not used anywhere in the code effectively)

[comment 3.] (https://github.com/cms-sw/cmssw/pull/39996#issuecomment-1348005385) I've updated the link and now it should be fine. Apologies.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-39996/33374

- This PR adds an extra 20KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-39996/33374/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-39996/33374/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

Cvico commented:
Hi, I'm not entirely sure why the code-checks fail now. I see this in the logs:
gmake[2]: Entering directory '/data/cmsbld/jenkins/workspace/run-pr-code-checks/CMSSW_13_0_X_2022-12-12-2300'

Although this PR was initially for CMSSW_12_6_0_pre4. Is there something I should do? 
Thanks.

perrotta commented:
@Cvico as the [message](https://github.com/cms-sw/cmssw/pull/39996#issuecomment-1348127177) says with the link below "code format", there is some indentation to fix in your code in order to comply to the cmssw standards.

Please do what suggested in the same [message](https://github.com/cms-sw/cmssw/pull/39996#issuecomment-1348127177), i.e.  run `scram build code-format` in your area, and commit and push the outcome obtained automatically with that `scram` command

Cvico committed:
Code-formats

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-39996/33377

- This PR adds an extra 32KB to repository

cmsbuild commented:
Pull request #39996 was updated. @SiewYan, @mkirsano, @Saptaparna, @cmsbuild, @alberto-sanchez, @menglu21, @GurpreetSinghChahal can you please check and sign again.

perrotta commented:
please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-f6f25e/29599/summary.html
**COMMIT**: bb0a15428d9dc56d357c5724a5a8ba181786384a
**CMSSW**: CMSSW_13_0_X_2022-12-12-2300/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/39996/29599/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 31 differences found in the comparisons
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3421510
 - DQMHistoTests: Total failures: 1176
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3420312
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 206 log files, 158 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_0_X_2022-12-12-2300+f6f25e/54649/triggerResults): no differences found

perrotta commented:
@cms-sw/generators-l2 is there anything that prevents you reviewing this PR? Could you please comment in any case, telling what you are waiting for, or when do you plan to review?

Saptaparna commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @perrotta, @dpiparo, @rappoccio (and backports should be raised in the release meeting by the corresponding L2)

perrotta commented on GeneratorInterface/Pythia8Interface/plugins/JetMatchingEWKFxFx.cc:
```diff
@@ -0,0 +1,998 @@
+
+int JetMatchingEWKFxFx::npNLO() {
+  string npIn = infoPtr->getEventAttribute("npNLO", true);
+  int np = (npIn.empty()) ? atoi((char*)npIn.c_str()) : -1;
```
One doubt of the last minute: is it correct that if `npIn` is empty you compute `npIn.c_str()`? It does not look so...

In any case `c_str()` returns a `const char*`, which is exaclty the argument needed by `std::atoi`: why do you need to cast a (non const) `char*`?

I think this should be (but please check):
```suggestion
  int np = !npIn.empty() ? atoi(npIn.c_str()) : -1;
```

perrotta commented on GeneratorInterface/Pythia8Interface/plugins/JetMatchingEWKFxFx.cc:
```diff
@@ -0,0 +1,998 @@
+  // Count the number of clusters with pT>qCut.  This includes the
+  //  original hard partons plus any hard emissions.
+  for (int idx = 0; idx < hjSlowJet->sizeAll(); ++idx) {
+    if (hjSlowJet->pT(idx) > sqrt(qCutSq))
```
```suggestion
    if (hjSlowJet->pT(idx) > qCut)
```
(or is it possible that `qCut` assumes negative values?)

Cvico replied:
In principle qCut should not take negative values so indeed this does can be done.

perrotta replied:
@Cvico this was not updated with the latest commit...

perrotta replied:
please commit now

perrotta commented on GeneratorInterface/Pythia8Interface/plugins/JetMatchingEWKFxFx.cc:
```diff
@@ -0,0 +1,998 @@
+  // Count the number of clusters with pT>qCut.  This includes the
+  //  original hard partons plus any hard emissions.
+  for (int idx = 0; idx < hjSlowJet->sizeAll(); ++idx) {
+    if (hjSlowJet->pT(idx) > sqrt(qCutSq))
```
```suggestion
    if (hjSlowJet->pT(idx) > qCut)
```
(or is it possible that `qCut` assumes negative values?)

Cvico replied:
In principle qCut should not take negative values so indeed this does can be done.

rappoccio commented:
hold

* Explicitly holding this until comments are addressed.

cmsbuild commented:
Pull request has been put on hold by  @rappoccio
They need to issue an `unhold` command to remove the `hold` state or L1 can `unhold` it for all

Cvico committed:
Fixed typos in some conditionals. Re run tests with the sample production and it's consistent with previous versions.

cmsbuild commented:
Pull request #39996 was updated. @SiewYan, @mkirsano, @Saptaparna, @cmsbuild, @alberto-sanchez, @menglu21, @GurpreetSinghChahal can you please check and sign again.

Cvico committed:
Addressed comment from reviewer. No need to recompute the sqrt(qCutSq) since the unsquared value (qCut) is already available

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-39996/33659

- This PR adds an extra 32KB to repository

cmsbuild commented:
Pull request #39996 was updated. @SiewYan, @mkirsano, @Saptaparna, @cmsbuild, @alberto-sanchez, @menglu21, @GurpreetSinghChahal can you please check and sign again.

perrotta commented:
please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-f6f25e/29916/summary.html
**COMMIT**: 6a878df4c89aa5487201f7dcd3d94e87c0b60b6d
**CMSSW**: CMSSW_13_0_X_2023-01-11-1100/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/39996/29916/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



There are some workflows for which there are errors in the baseline:
 140.01 step 3
140.03 step 3
 The results for the comparisons for these workflows could be incomplete 
This means most likely that the IB is having errors in the relvals.The error does NOT come from this pull request

**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 1 differences found in the comparisons
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3555538
 - DQMHistoTests: Total failures: 3
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3555513
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 48 files compared)
 - Checked 211 log files, 162 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_0_X_2023-01-11-1100+f6f25e/54940/triggerResults): no differences found

perrotta commented:
unhold

Cvico commented:
Dear reviewers,

I'd like to thank you for the thorough review of my the plugin, and also to ask you
if there's anything more I can do to proceed with the next steps for the final inclusion
of the PR?

Best,
Carlos Vico

perrotta commented:
> Dear reviewers,
> 
> I'd like to thank you for the thorough review of my the plugin, and also to ask you if there's anything more I can do to proceed with the next steps for the final inclusion of the PR?
> 
> Best, Carlos Vico

@Cvico just waiting for the @cms-sw/generators-l2 signature (the same as in quite several other PRs which just miss that one...)

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-f6f25e/30325/summary.html
**COMMIT**: 6a878df4c89aa5487201f7dcd3d94e87c0b60b6d
**CMSSW**: CMSSW_13_0_X_2023-01-31-2300/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/39996/30325/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially added 3 lines to the logs
 - Reco comparison results: 6 differences found in the comparisons
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3555495
 - DQMHistoTests: Total failures: 3
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3555470
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 48 files compared)
 - Checked 211 log files, 162 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_0_X_2023-01-31-2300+f6f25e/55280/triggerResults): no differences found

Saptaparna commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @perrotta, @dpiparo, @rappoccio (and backports should be raised in the release meeting by the corresponding L2)

perrotta commented:
+1

Cvico commented:
Hi,
Thank you all for reviewing and finally accepting my PR!

I'd like this to be backported to CMSSW_12_4_X in order to use this in Run3 production. How should
I proceed in order to do this? Thanks in advance for the feedback.

perrotta commented:
@Cvico please organize with @cms-sw/generators-l2 possible backport strategies of this PR, starting with deciding whether really a production with 12_4_X is what you need now, or if you should point at some more recent release instead
